renovate/src/Renovate/BasicBlock/Types.hs

```diff
@@ -171,17 +233,31 @@ type SymbolicBlock arch = BasicBlock (SymbolicInfo arch) (TaggedInstruction arch
 --
 -- We use the parameter to choose between symbolic and concrete
 -- addresses for the successors.
-data FallthroughTag a = FallthroughTag
-  { ftTarget :: Maybe a
-  , ftFallthrough :: Maybe a
-  } deriving (Eq, Ord, Read, Show, Functor, Foldable, Traversable)
+data FallthroughType a where
+  -- | A tag for instructions internal to a block that implicitly fall through
+  -- to the next instruction but need no correction
+  InternalInstruction :: FallthroughType a
+  -- | The end of a basic block that unconditionally allows execution to fall
+  -- through to the next block
+  UnconditionalFallthrough :: a -> FallthroughType a
+  -- | The block ends in an unconditional jump to the named target
+  UnconditionalJump :: a -> FallthroughType a
+  -- | The block ends in a conditional jump, and thus has two targets (in
+  -- order): the target of the conditional branch and the implicit fallthrough
+  -- target
+  ConditionalFallthrough :: a -> a -> FallthroughType a
+
+deriving instance Functor FallthroughType
+deriving instance Foldable FallthroughType
+deriving instance Traversable FallthroughType
+deriving instance (Show a) => Show (FallthroughType a)
 
 -- | A wrapper around a normal instruction that includes optional addresses for
 -- all of the instruction's successors that may be relocated. The addresses can
 -- be chosen to be symbolic or concrete by picking different @addr@s.
 data FallthroughInstruction arch addr a = FallthroughInstruction
   { ftInstruction :: Instruction arch a
```

Comment:
`ftInstruction` => `fallthroughInsn` (or similar) to be consistent with the other record field name?

Reply:
It _seems_ (from my `c+f`-ing the diff view here)`FallthroughInstruction` is _only_ ever used with `SymbolicAddress`'s -- if that is the case it would be nice to drop the `addr` param (and if it's not true, constrain the `arch` and `addr` params to relate or have separate types...?):

```
data FallthroughInstruction arch a = FallthroughInstruction
{ fallthroughInsn :: Instruction (SymbolicAddress arch) a
  fallthroughType :: FallthroughType (SymbolicAddress arch)
}
```

Reply:
The `ConcreteFallthrough` and `SymbolicFallthrough` types are used in a few places (they are wrappers around `FallthroughInstruction`